test/source_test.go

```diff
@@ -1,20 +1,338 @@
 package test
 
 import (
+	"bufio"
+	"bytes"
 	"context"
+	"fmt"
 	"io/fs"
+	"net"
+	"net/http"
 	"os"
+	"os/user"
 	"path/filepath"
 	"strings"
+	"sync"
 	"testing"
 
+	githttp "github.com/AaronO/go-git-http"
+	"github.com/AaronO/go-git-http/auth"
 	"github.com/Azure/dalec"
 	"github.com/Azure/dalec/frontend/pkg/bkfs"
+	"github.com/Azure/dalec/test/testenv"
+	ps "github.com/mitchellh/go-ps"
 	"github.com/moby/buildkit/client/llb"
 	gwclient "github.com/moby/buildkit/frontend/gateway/client"
 	"github.com/opencontainers/go-digest"
 )
 
+var (
+	isRootLess     bool
+	isRootlessOnce sync.Once
+)
+
+func TestGomodGitAuthHTTPS(t *testing.T) {
+	t.Parallel()
+
+	// this is needed because go behaves differently when the host is
+	// recognized if it sees `host.docker.internal` it will first try to
+	// resolve the metadata by connecting to the resolved IP address on port
+	// 443, which will fail.
+	const host = "host.docker.internal"
+
+	ctx := startTestSpan(baseCtx, t)
+	sourceName := "gitauth"
+
+	// randomData := identity.NewID()
+
+	testEnv.RunTest(ctx, t, func(ctx context.Context, c gwclient.Client) {
+		const gomodFmt = `module %[1]s/user/public
+
+go 1.23.5
+
+require %[1]s/user/private.git v0.0.0
+`
+
+		gomodContents := fmt.Sprintf(gomodFmt, host)
+		port := getAvailablePort(t)
+
+		spec := &dalec.Spec{
+			Name: "gomod-git-auth",
+			Sources: map[string]dalec.Source{
+				sourceName: {
+					Inline: &dalec.SourceInline{
+						Dir: &dalec.SourceInlineDir{
+							Files: map[string]*dalec.SourceInlineFile{
+								"go.mod": {
+									Contents: gomodContents,
+								},
+							},
+						},
+					},
+					Generate: []*dalec.SourceGenerator{
+						{
+							Gomod: &dalec.GeneratorGomod{
+								Auth: map[string]dalec.GomodGitAuth{
+									fmt.Sprintf("%s:%d", host, port): {
+										Token: "super-secret",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		}
+
+		outsideAddr, insideAddr := getGitServerAddrs(ctx, t, c)
```

Comment:
I don't think we are using these values anywhere other than `runGitServer`, maybe it makes sense to have `runGitServer` handle all the things here?

Reply:
`insideAddr` gets added as an extra host a few lines down.